reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2020-Present Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.core.publisher;
+
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Function;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.FluxConcatMap.ConcatMapInner;
+import reactor.core.publisher.FluxConcatMap.FluxConcatMapSupport;
+import reactor.core.publisher.FluxConcatMap.WeakScalarSubscription;
+import reactor.util.annotation.Nullable;
+import reactor.util.context.Context;
+
+/**
+ * Maps each upstream value into a Publisher and concatenates them into one
+ * sequence of items.
+ *
+ * @param <T> the source value type
+ * @param <R> the output value type
+ *
+ * @see FluxConcatMap
+ */
+final class FluxConcatMapNoPrefetch<T, R> extends InternalFluxOperator<T, R> {
+
+	final Function<? super T, ? extends Publisher<? extends R>> mapper;
+
+	FluxConcatMapNoPrefetch(
+			Flux<? extends T> source,
+			Function<? super T, ? extends Publisher<? extends R>> mapper
+	) {
+		super(source);
+		this.mapper = Objects.requireNonNull(mapper, "mapper");
+	}
+
+	@Override
+	public CoreSubscriber<? super T> subscribeOrReturn(CoreSubscriber<? super R> actual) {
+		if (FluxFlatMap.trySubscribeScalarMap(source, actual, mapper, false, true)) {
+			return null;
+		}
+
+		return new FluxConcatMapNoPrefetchSubscriber<>(actual, mapper);
+	}
+
+	static final class FluxConcatMapNoPrefetchSubscriber<T, R> implements FluxConcatMapSupport<T, R> {
+
+		enum State {
```

Comment:
Could you document the transition diagram, or at least explain the states in more detail (in particular LAST_ACTIVE)?

Reply:
Unless you have a good idea how to _code_ it, I don't think doing a diagram is a good idea, since it may easily get outdated :(

Reply:
I meant at least documenting what transitions make sense, and what each (non trivial) state represents. That said, I would argue that if we were to produce a diagram for comprehension's sake and it got outdated, there is indeed a problem :)

Reply:
I believe that code is the best documentation especially when state transition is not that complicated and can be easily derived from the enums.